web/html/src/components/package/PackageListActionScheduler.tsx

```diff
@@ -0,0 +1,256 @@
+  actionType: string;
+  listTitle: string;
+  listSummary: string;
+  listEmptyText: string;
+  listActionLabel: string;
+  listColumns: React.ReactNode[];
+  confirmTitle: string;
+};
+
+type State = {
+  messages: Array<MessageType>;
+  selectedPackages: string[];
+  confirmAction: boolean;
+  earliest: moment.Moment;
+  actionChain?: ActionChain;
+};
+
+export class PackageListActionScheduler extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      messages: [],
+      selectedPackages: [],
+      confirmAction: false,
+      earliest: localizedMoment(),
+    };
+  }
+
+  confirmAction = () => {
+    this.setState({
+      confirmAction: true,
+      messages: [],
+    });
+  };
+
+  backToList = () => {
+    this.setState({
+      confirmAction: false,
+      messages: [],
+    });
+  };
+
+  scheduleAction = () => {
+    const requestBody = {
+      actionType: this.props.actionType,
+      earliest: this.state.earliest,
+      actionChain: this.state.actionChain ? this.state.actionChain.text : null,
+      selectedPackages: this.state.selectedPackages,
+    };
+
+    Network.post(this.props.scheduleActionAPI, requestBody)
+      .then((data) => {
+        // Notify the successful outcome
+        const msg = MessagesUtils.info(
+          this.state.actionChain ? (
+            <span>
+              {t("Action has been successfully added to the ")}
+              <ActionChainLink id={data}>{this.state.actionChain.text}</ActionChainLink>
+            </span>
+          ) : (
+            <span>
+              {t("The action has been ")}
+              <ActionLink id={data}>{t("scheduled.")}</ActionLink>
+            </span>
+          )
+        );
+
+        // Clear the current selection
+        Network.post(`/rhn/manager/api/sets/${this.props.selectionSet}/clear`).catch((err) => {
+          this.setState({ messages: MessagesUtils.warning(t("Unable to clear selection")) });
+        });
+
+        this.setState({
+          confirmAction: false,
+          messages: msg,
+        });
+      })
+      .catch((err) => {
+        this.setState({ messages: MessagesUtils.error(t("Unable to perform action.")) });
+      });
+  };
+
+  onDateTimeChanged = (date) => {
+    this.setState({ earliest: date });
+  };
+
+  onActionChainChanged = (actionChain) => {
+    this.setState({ actionChain: actionChain });
+  };
+
+  searchData = (item, criteria) => {
+    if (criteria) {
+      const value = this.packageNameFromSelectionKey(item);
+      return value.toLocaleLowerCase().includes(criteria.toLocaleLowerCase());
+    }
+
+    return true;
+  };
+
+  packageNameLink = (item) => {
+    if (item.packageId == null) {
+      return item.nvre;
+    }
+
+    return <a href={`/rhn/software/packages/Details.do?pid=${item.packageId}`}>{item.nvre}</a>;
+  };
+
+  buildSelectionKey = (item) => {
+    if (item.nvrea != null) {
+      return item.idCombo + "~*~" + item.nvrea;
```

Comment:
If `"~*~"` is some kind of a special token we want to use in multiple places, I'd recommend lifting it out into a const variable e.g. `const FOO_TOKEN = "~*~";`.

Reply:
Good point. It's indeed how selection key is generated on the server side (see the java class `PackageListItem` methods `getSelectionKey()` and `parse()`). I kept this mechanism because it's how it works for the existing (struts) pages that deals with packages. In this way, it should be fairly easy to reuse this component unchanged to convert them to React.

Reply:
I extracted the constant